### ISSUE TITLE
Template::custom but without configured template directory

### ISSUE THREAD

**Issue by ssendev** [SergioBenitez/Rocket#1792]
**Existing Functionality**

I want to ship a single file without a templates directory.

One way it is possible now is to use a managed state Tera instance but this requires changing the routes which means it is not possible to have templates included in release and reloaded from files during development.

With `rocket.attach(Template::custom(…` it is possible to `cfg!(debug_assertions)`   `engines.tera.add_raw_template("index", include_str!("../templates/index.html.tera"))`. This works but still requires an empty templates folder to be present.

**Suggested Changes**

Add `Template::fully_custom` which uses `Tera::default` instead of `Tera::new`

**Alternatives Considered**

- Adding `tera.clear_config()` which would reset the Tera instance to its `Tera::defalt()` state and could be called in `Template::custom`

- Ideally there would be a `Template::compiled_in_fairing()` which would behave like `Template::fairing` but load templates at compile time but this might be hard to implement.

**Comment by jebrosen** [comment]
> I want to ship a single file without a templates directory.
> 
> One way it is possible now is to use a managed state Tera instance but this requires changing the routes which means it is not possible to have templates included in release and reloaded from files during development.

This actually seems *pretty* close to a good solution already: you could manually manage a `Tera` instance with the `include_str` to cover release builds, and in debug only, load the template from the file at every render instead of using the compiled-in copy.

> 
> With `rocket.attach(Template::custom(…` it is possible to `cfg!(debug_assertions)` `engines.tera.add_raw_template("index.html", include_str!("../templates/index.html.tera"))`. This works but still requires an empty templates folder to be present.
> 
> (Well it almost works. Since rocket strips two extensions and templates added with `add_raw_template` only one template names aren't the same in debug and release. Fortunaltely it is possible to create symlinks from `index.html.html.tera` to `index.html.tera` and then one can use `index.html` as template name)

Note that you could also call `engines.tera.add_raw_template("index", include_str!("../templates/index.html.tera"))` in order to use the same name in both cases, instead of the symlink.

> **Suggested Changes**
> 
> Add `Template::fully_custom` which uses `Tera::default` instead of `Tera::new`

`Tera::default` is already used: <https://api.rocket.rs/v0.5-rc/src/rocket_dyn_templates/tera_templates.rs.html#15>. I think the desired effect here is simply for Rocket to (optionally) not load any templates, and additionally for the configuration to not require a template directory. But yes, I think it would be possible to implement this.

> * Ideally there would be a `Template::compiled_in_fairing()` which would behave like `Template::fairing` but load templates at compile time but this might be hard to implement.

Yes - unfortunately this option is a *lot* of complexity to add especially at compile time. But I think this need should already be largely fulfilled by other compile-time template engines.

**Comment by ssendev** [comment]
> This actually seems _pretty_ close to a good solution already: you could manually manage a `Tera` instance with the `include_str` to cover release builds, and in debug only, load the template from the file at every render instead of using the compiled-in copy.

That could work but would still require returning something else than `Template`. Or maybe implement some trait to make it work?

> Note that you could also call `engines.tera.add_raw_template("index", include_str!("../templates/index.html.tera"))` in order to use the same name in both cases, instead of the symlink.

That's what I tried first and thought worked until i realized that they were served with mime type `text/plain` maybe there is another way to get the right mime type without the double html.

> `Tera::default` is already used: https://api.rocket.rs/v0.5-rc/src/rocket_dyn_templates/tera_templates.rs.html#15. I think the desired effect here is simply for Rocket to (optionally) not load any templates, and additionally for the configuration to not require a template directory. But yes, I think it would be possible to implement this.

Yes I didn't realize rocket was already loading the templates itself I just found the Tera docs that said `new` loads templates `default` does not. I just tried copying the contents of that file to my crate but couldn't proceed since `rocket_dyn_templates::engine` is private. But even if that worked I may have kept the version with the symlinks since that solution requires less code.

**Comment by jebrosen** [comment]
Yes, that particluar suggestion includes to not using `Template` at all: you would instead call `.render()`, possibly via your own `Responder` implementation.

And ouch, I neglected to consider the content-type. The symlink solution seems appropriate since it's already working as desired, and I think it's unlikely for that approach to fundamentally change.

**Comment by cubetastic33** [comment]
Do you know if it's possible to configure `template_dir` during compile time? I was able to get it working without an empty templates directory if I set `template_dir` to `"."`, but I needed Rocket.toml to configure that and it didn't work when I ran the binary from a different directory. The [rocket::Config](https://api.rocket.rs/v0.5-rc/rocket/config/struct.Config.html) struct doesn't have an option to set this value either.

**Comment by jebrosen** [comment]
Sure! Values can be [set programmatically](https://rocket.rs/v0.5-rc/guide/configuration/#custom-providers) and used with the the `rocket::custom` constructor.

**Comment by ogarcia** [comment]
I think there should be an option similar to `embed_migrations!()` which now exists for databases. I try with [rust-embed](https://crates.io/crates/rust-embed) but is more complicated than `add_raw_template` and the problem is the same.

**Comment by ogarcia** [comment]
Another problem I have recently encountered with templates when trying to use `add_raw_template` is that if you use several templates and reference one to another it does not work as it should.

<details>
  <summary>For example this code:</summary>

```rust
#[macro_use] extern crate rocket;

use rocket_dyn_templates::{context, Template};

#[get("/")]
fn index() -> Template {
    Template::render("index", context! {
        foo: 123,
    })
}

#[get("/other")]
fn other() -> Template {
    Template::render("other", context!{
        foo: 321,
    })
}

#[launch]
fn rocket() -> _ {
    rocket::build()
        .attach(Template::custom(|engines| {
            let mut templates = Vec::new();
            templates.push(("index.html", "{% block other %}{% endblock %} Foo: {{ foo }}"));
            templates.push(("other.html", "{% extends \"index\" %}{% block other %}Hello{% endblock %}"));
            engines.tera.add_raw_templates(templates).unwrap();
        }))
        .mount("/", routes![index, other])
}
```

</details>

What I do is to add two simple templates that reference each other (`{% extends \"index\" %}`) and what happens is that when launching the application gives this error:
```
thread 'main' panicked at src/main.rs:24:55:
called `Result::unwrap()` on an `Err` value: Error { kind: MissingParent { current: "other.html", parent: "index" }, source: None }
```

The only workaround for the code to work is to remove the `.html` from the name of the templates, but doing so renders the templates as text instead of html.

**Comment by the10thWiz** [comment]
@ogarcia In your example, does changing the name in the extends to `{% extends \"index.html\" %}` work?

**Comment by the10thWiz** [comment]
A potentially better solution would be to add a `register_templates(tms: impl IntoIter<RawTemplate>)` method to the generated Fairing, which does the following - disables the error for a missing template dir, and registers the raw templates listed by the iterator. The `RawTemplate` struct would contain an engine identifier (i.e. Tera vs Handlebars, etc), a Content-Type, a name, and the raw text of the template.

There could be additional logic/configuration to make this easier to use, but I think that will pop out once an implementation has been created.

**Comment by ogarcia** [comment]
> [@ogarcia](https://github.com/ogarcia) In your example, does changing the name in the extends to `{% extends \"index.html\" %}` work?

No, it starts but then gives you a 500 saying that the index template does not exist.
```
GET / text/html:
   >> Matched: (index) GET /
   >> Template 'index' does not exist.
   >> Known templates: index.html, other.html.
   >> Searched in "/rust/mycode/templates".
   >> Outcome: Error(500 Internal Server Error)
   >> No 500 catcher registered. Using Rocket default.
   >> Response succeeded.
```

> The `RawTemplate` struct would contain an engine identifier (i.e. Tera vs Handlebars, etc), a Content-Type, a name, and the raw text of the template.

Yes, the idea of being able to indicate the type of template, the content-type and the template in raw seems to me of the most successful. If we add to that the fact that the `templates` directory does not need to exist, everything would be solved.
